Fix flaky crash in ResourceMonitor.

basuke opened the pull request:
#### 734c49b3d075dcc33f56becf3bde8aca5245b719
<pre>
Fix flaky crash in ResourceMonitor.
<a href="https://bugs.webkit.org/show_bug.cgi?id=288357">https://bugs.webkit.org/show_bug.cgi?id=288357</a>
<a href="https://rdar.apple.com/145427232">rdar://145427232</a>

Reviewed by Chris Dumez.

In ResourceMonitor.cpp, a macro to produce RELEASE_LOG is defined and m_frame weak ptr is used without checking.

Add check before calling that macro.

* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::setEligibility):
(WebCore::ResourceMonitor::addNetworkUsage):
(WebCore::ResourceMonitor::updateNetworkUsageThreshold):
* Source/WebCore/loader/ResourceMonitorChecker.cpp:
(WebCore::ResourceMonitorChecker::setNetworkUsageThreshold):

Canonical link: <a href="https://commits.webkit.org/290954@main">https://commits.webkit.org/290954@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b25318317a74ca036fd8d0362b7bf79fad50626

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91558 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11089 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/622 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96527 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42244 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93608 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11467 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19515 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/70310 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/27831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94559 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/8759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82952 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50634 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/8522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/541 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41414 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/78842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/544 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98532 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18719 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79335 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18974 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78790 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78538 "Found 1 new API test failure: /WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19425 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23061 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/417 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18715 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23991 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18425 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21886 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20192 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -65,6 +65,10 @@ ResourceMonitor::~ResourceMonitor()
 
 void ResourceMonitor::setEligibility(Eligibility eligibility)
 {
+    RefPtr frame = m_frame.get();
```
Seems wrong to store m_frame in a RefPtr just to null check it..

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -146,6 +150,11 @@ void ResourceMonitor::addNetworkUsage(size_t bytes)
     if (m_networkUsageExceed)
         return;
 
+    RefPtr frame = m_frame.get();
```
Seems wrong to store m_frame in a RefPtr just to null check it..

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -159,6 +168,10 @@ void ResourceMonitor::updateNetworkUsageThreshold(size_t threshold)
     if (m_networkUsageThreshold == threshold)
         return;
 
+    RefPtr frame = m_frame.get();
```
Seems wrong to store m_frame in a RefPtr just to null check it..

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -159,6 +168,10 @@ void ResourceMonitor::updateNetworkUsageThreshold(size_t threshold)
+    if (!frame)
+        return;
+
     RESOURCEMONITOR_RELEASE_LOG("Update network usage threshold: threshold=%zu", threshold);
```
The fact that `RESOURCEMONITOR_RELEASE_LOG` uses `m_frame` and that the call site has to know to null check it is not obvious at all. This is bad design that will cause more bugs in the future. I think we need to update `RESOURCEMONITOR_RELEASE_LOG` to null check m_frame internally.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -41,7 +41,8 @@ namespace WebCore {
 
-#define RESOURCEMONITOR_RELEASE_LOG(fmt, ...) RELEASE_LOG(ResourceMonitoring, "ResourceMonitor(frame %" PRIu64 ")::" fmt, m_frame->frameID().object().toUInt64(), ##__VA_ARGS__)
+#define RESOURCEMONITOR_RELEASE_LOG(fmt, ...) if (m_frame) \
+    RELEASE_LOG(ResourceMonitoring, "ResourceMonitor(frame %" PRIu64 ")::" fmt, m_frame->frameID().object().toUInt64(), ##__VA_ARGS__)
```
I think this can use `RELEASE_LOG_IF(m_frame, ...)` instead of doing an explicit `if` check

cdumez reviewed (approved):
LGTM with suggestion.

basuke committed:
Fix flaky crash in ResourceMonitor.
https://bugs.webkit.org/show_bug.cgi?id=288357
rdar://145427232

Reviewed by Chris Dumez.

In ResourceMonitor.cpp, a macro to produce RELEASE_LOG is defined and m_frame weak ptr is used without checking.

Add check before calling that macro.

* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::setEligibility):
(WebCore::ResourceMonitor::addNetworkUsage):
(WebCore::ResourceMonitor::updateNetworkUsageThreshold):
* Source/WebCore/loader/ResourceMonitorChecker.cpp:
(WebCore::ResourceMonitorChecker::setNetworkUsageThreshold):

Canonical link: https://commits.webkit.org/290954@main
